File: pyround1/290.单词规律.py
#
# @lc app=leetcode.cn id=290 lang=python
#
# [290] 单词规律
#

# @lc code=start
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        lookup1 = {}
        lookup2 = {}

        wordList = str.split(' ', -1) 

        if len(pattern) != len(wordList):
            return False
        
        for i in range(len(pattern)):
            if lookup1.get(pattern[i]):
                if lookup1[pattern[i]] != wordList[i]:
                    return False
            else:
                lookup1[pattern[i]] = wordList[i]

            if lookup2.get(wordList[i]):
                if lookup2[wordList[i]] != pattern[i]:
                    return False
            else:
                lookup2[wordList[i]] = pattern[i]

        return True
# @lc code=end

File: 290.单词规律.cpp
/*
 * @lc app=leetcode.cn id=290 lang=cpp
 *
 * [290] 单词规律
 */

// @lc code=start
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<string, char> res;
        string tmp = "";
        for(int i = 0; i < s.size(); i++){
            if(s[i] != ' ')
                tmp += s[i];
            else{
                if(res[tmp])
                {
                    if(res[tmp] != tmp)
                        return false;
                }
                else{
                    res[tmp] = tmp;
                }
                tmp = "";
            }
        }
        return true;

    }
};
// @lc code=end

